issue-250-ridge-transforms

Hojat-Shirmard opened the pull request:
Search for "hjt" in the file; you'll see I've added lines of code in four different places.

michaelchin commented:
@Hojat-Shirmard 

- I have changed the title to "issue-250-ridge-transforms". It is more descriptive. 
- Can you move the two functions to the end of the file and fix the indentation?
- let's add "ridge_transforms" as property(return self.ridges + self.transforms), similar to this https://github.com/GPlates/gplately/blob/bea66aaff046fa7d08e33c94c41b63695f8ecbe9/gplately/plot.py#L383-L387
- create a test program, similar to https://github.com/GPlates/gplately/blob/issue-250-ridge-transforms/tests-dir/unittest/test_plot_with_raster.py (remove the parts which you don't need.)

Hojat-Shirmard commented:
Hi @michaelchin 
I made a second commit considering your requested changes. To test the code, should I add `plot.plot.ridge_transforms` to the code you shared and run it in the Jupyter notebook?

michaelchin commented:
> Hi @michaelchin I made a second commit considering your requested changes. To test the code, should I add `plot.plot.ridge_transforms` to the code you shared and run it in the Jupyter notebook?

yes, add plot_ridge_transform() funciton call and comment out other plotting code to see the result more clearly.

And you  can see your code changes here https://github.com/GPlates/gplately/pull/298/files. there are still errors in the code changes.

Hojat-Shirmard commented:
How can I know where the errors are?

michaelchin commented:
> How can I know where the errors are?

Have you tried to click this link https://github.com/GPlates/gplately/pull/298/files?

if you did, see the screenshot below for example. 

You need to make sure the changes are intended. And only make changes when it is necessary.

<img width="1460" alt="Screenshot 2024-11-02 at 3 56 22 PM" src="https://github.com/user-attachments/assets/243728e3-39c6-4b9c-89dc-6b03aeffdcf0">

Hojat-Shirmard commented:
I reckon I need to create a new environment in Anaconda and install my modified branch to check if it’s working properly, right? 
I've also passed some courses on YouTube and learned a bit about Docker, so if there are any instructions or steps to follow with Docker, I’d be happy to try that as well.

Hojat-Shirmard commented:
I added logger.debug
I added the "ridge_transforms" attribute, but the result is not correct!
![plot](https://github.com/user-attachments/assets/34ac900f-32e4-4f91-9da1-2225204090c9)

jcannon-gplates commented:
Maybe it's because the `@property` decorator is missing. See Michael's example above...

```
 @property 
 def topological_plate_boundaries(self): 
     if self._topological_plate_boundaries is None: 
         self.update_time(self.time) 
     return self._topological_plate_boundaries
```

You can see [here](https://realpython.com/python-property/) how properties are added.

Hojat-Shirmard commented:
Hi @jcannon-gplates 
Thank you for your quick response!
I just fixed the property, and here is the result:
![plot2](https://github.com/user-attachments/assets/4e8dd2e8-c715-4ad8-a737-1894773381c8)

jcannon-gplates commented:
It looks like you're using `self._topological_plate_boundaries`. That will work since it'll generate `self.ridges` and `self.transforms` if those attributes don't exist yet.

However, it might be better to do something like...

```
    @property
    def ridge_transforms(self):
        """Combine ridges and transforms into one feature set."""
        if self.time is None:
            raise ValueError(
                "No topologies have been resolved. Set `PlotTopologies.time` to construct them."
            )
        logger.debug("Returning ridge_transforms property.")
        return self.ridges + self.transforms
```

...where I've replaced...

```
        if self._topological_plate_boundaries is None:
            logger.debug("Accessing ridge transforms - updating topological plate boundaries.")
            self.update_time(self.time)
```

...with...

```
        if self.time is None:
            raise ValueError(
                "No topologies have been resolved. Set `PlotTopologies.time` to construct them."
            )
```

...because if the time hasn't been set (in either `__init__` or with `self.time = ...`), and hence `self.time` (as well as `self._topological_plate_boundaries`) is `None`, then `self.update_time(self.time)` will fail [here](https://github.com/GPlates/gplately/blob/a61e9237d19158c5071d22ce4882db465bb4e045/gplately/plot.py#L452).

@GPlates/gplately-dev, does this sound right? I think `self.update_time()` probably should only be called on setters (like `self.anchor_plate_id = ...`). Getters should already be up-to-date (since `update_time()` has already been called behind the scenes whenever the class state changed, via setters, etc). Getters (that rely on variables updated by `update_time()`) should perhaps instead raise an error (if the time has not been set yet). My understanding is the user should eventually set the time if they want to access certain functionalities of the class. For example, `self.ridges` and `self.transforms` don't exist if the user has not yet set the time, so an error should be raised if they try to access them.

Hojat-Shirmard commented:
Thank you so much! I just applied your comment.

michaelchin commented:
Thanks @jcannon-gplates and @Hojat-Shirmard 

Just out of curiosity, what was the reason of these two lines got removed?

<img width="1901" alt="Screenshot 2024-11-26 at 5 08 13 pm" src="https://github.com/user-attachments/assets/528c53ef-66e4-46ea-89ba-2fbd05963353">

Hojat-Shirmard commented:
I think I messed up—sorry about that! I'll fix it now.

michaelchin commented:
@Hojat-Shirmard Can you add other properties and functions which were mentioned in https://github.com/GPlates/gplately/issues/250? And also add the warning and debug messages mentioned in https://github.com/GPlates/gplately/issues/250. Thanks.

Hojat-Shirmard commented:
@michaelchin, Did you mean this comment?
https://github.com/GPlates/gplately/issues/250#issuecomment-2275012714

Hojat-Shirmard commented:
I added misc_transforms
